upgrade_tests/upgrade_through_versions_test.py

```diff
@@ -942,18 +947,20 @@ def create_upgrade_class(clsname, version_metas, protocol_version,
 for upgrade in MULTI_UPGRADES:
     # if any version_metas are None, this means they are versions not to be tested currently
     if all(upgrade.version_metas):
-        metas = upgrade.version_metas
+        # even for RUN_STATIC_UPGRADE_MATRIX we only test upgrade paths jdk compatible with the end "indev_" version
+        metas = jdk_compatible_steps(upgrade.version_metas)
 
         if not RUN_STATIC_UPGRADE_MATRIX:
-            if metas[-1].matches_current_env_version_family:
-                # looks like this test should actually run in the current env, so let's set the final version to match the env exactly
-                oldmeta = metas[-1]
-                newmeta = oldmeta.clone_with_local_env_version()
-                logger.debug("{} appears applicable to current env. Overriding final test version from {} to {}".format(upgrade.name, oldmeta.version, newmeta.version))
-                metas[-1] = newmeta
-                create_upgrade_class(upgrade.name, [m for m in metas], protocol_version=upgrade.protocol_version, extra_config=upgrade.extra_config)
-        else:
-            create_upgrade_class(upgrade.name, [m for m in metas], protocol_version=upgrade.protocol_version, extra_config=upgrade.extra_config)
+            # replace matching meta with current version
+            for idx, meta in enumerate(metas):
+                if meta.matches_current_env_version_family:
+                    java_version = current_env_java_version()
```

Comment:
we should get that java version once and do not do that in every iteration in a loop

Reply:
it should only called once, i was missing a `break`